stablehlo/reference/Ops.cpp

```diff
@@ -314,6 +325,147 @@ Tensor evalRealOp(const Tensor &operand, TensorType resultType) {
+      arg.set({}, runtimeResult.get(resultIndex));
+      args.push_back(arg);
+    }
+    for (auto [input, initValue] : llvm::zip(inputs, initValues)) {
+      auto arg = Tensor(initValue.getType());
+      arg.set({}, input.get(*inputIt));
+      args.push_back(arg);
+    }
+    SmallVector<Tensor> reducedValues = eval(body, args, &scope);
+    for (auto [result, value] : llvm::zip(results, reducedValues))
+      result.set(resultIndex, value.get({}));
+  }
+  return results;
+}
+
+SmallVector<Tensor> evalReduceWindowOp(
+    ArrayRef<Tensor> inputs, ArrayRef<Tensor> initValues,
+    const Axes &windowDimensions, const Axes &windowStrides,
+    const Axes &baseDilations, const Axes &windowDilations,
+    const SmallVector<SmallVector<int64_t>> &padding, Region &body,
+    Scope &scope, ArrayRef<TensorType> resultTypes) {
+  SmallVector<Tensor> runtimeResults;
+  Builder builder = mlir::Builder(resultTypes[0].getContext());
+  Type i64Type = builder.getI64Type();
+
+  for (auto [resultType, initValue] : llvm::zip(resultTypes, initValues)) {
+    Tensor result = Tensor(resultType);
+    for (auto resultIt = result.index_begin(); resultIt != result.index_end();
+         ++resultIt)
+      result.set(*resultIt, initValue.get({}));
+    runtimeResults.push_back(result);
+  }
+
+  SmallVector<int64_t> paddingLow;
+  SmallVector<int64_t> paddingHigh;
+  for (auto paddingPair : padding) {
+    paddingLow.push_back(paddingPair[0]);
+    paddingHigh.push_back(paddingPair[1]);
+  }
+
+  SmallVector<Tensor> paddedInputs;
+  for (auto [input, initValue] : llvm::zip(inputs, initValues)) {
+    SmallVector<Type> inferredPadType;
+    auto padStatus = hlo::inferPadOp(
+        {}, input.getType(), initValue.getType(),
+        getDenseIntElementsAttr({static_cast<int64_t>(paddingLow.size())},
+                                i64Type, paddingLow),
+        getDenseIntElementsAttr({static_cast<int64_t>(paddingHigh.size())},
+                                i64Type, paddingHigh),
+        getDenseIntElementsAttr({static_cast<int64_t>(baseDilations.size())},
+                                i64Type, llvm::to_vector(baseDilations)),
+        inferredPadType);
+    if (failed(padStatus))
+      report_fatal_error(
+          invalidArgument("Could not infer PadOp's return type"));
+    auto paddedInput =
+        evalPadOp(input, initValue, Sizes(paddingLow), Sizes(baseDilations) - 1,
+                  inferredPadType[0].cast<TensorType>());
+    paddedInputs.push_back(paddedInput);
+  }
+
+  for (auto resultIt = runtimeResults[0].index_begin();
+       resultIt != runtimeResults[0].index_end(); ++resultIt) {
+    SmallVector<Tensor> windows;
+    auto windowStart = (*resultIt) * Sizes(windowStrides);
+    for (auto paddedInput : paddedInputs) {
+      SmallVector<int64_t> limitIndices;
+      for (size_t i = 0; i < windowStart.size(); ++i)
+        limitIndices.push_back(
+            std::min(windowStart[i] + windowDimensions[i] * baseDilations[i],
```

Comment:
```
(C5) dim(result, d) =  (limit_indices[d]-start_indices[d])/stride[d] for all dimension d in operand.
 ```
I think we do need it. The above is the constraint in SliceOp, and if limit_indices do not account for strides, the shape inference would be wrong. Here is an example slice op:
```
func.func @slice_op() {
  %operand = stablehlo.constant dense<[[1, 1, 1],
                                       [1, 1, 1],
                                       [1, 1, 1]]> : tensor<3x6xi64>
  %result = "stablehlo.slice"(%operand) {
    start_indices = dense<[0, 0]> : tensor<2xi64>,
    limit_indices = dense<[3, 3]> : tensor<2xi64>,
    strides = dense<[10, 10]> : tensor<2xi64>
  } : (tensor<3x6xi64>) -> tensor<1x1xi64>
  check.expect_eq_const %result, dense<[[1]]> : tensor<1x1xi64>
  func.return
}
```

Reply:
I had the same concern you said, but here is the thing. 

While we are interpreting a StableHLO program, we are assuming that the program is valid. The purpose of calling inferSlice from a `evalOp` is not to check the validity of the result type but to just get the result type so that we can invoke `evalSliceOp`  and while doing that we can take liberty on the assumption that the program is valid.

Reply:
I see what you mean, thanks for the clarification!